adapters/repos/db/queue/scheduler.go

```diff
@@ -0,0 +1,490 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package queue
+
+import (
+	"context"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/adapters/repos/db/vector/common"
+	enterrors "github.com/weaviate/weaviate/entities/errors"
+)
+
+type Scheduler struct {
+	SchedulerOptions
+
+	queues struct {
+		sync.Mutex
+
+		m map[string]*queueState
+	}
+
+	// context used to close pending tasks
+	ctx      context.Context
+	cancelFn context.CancelFunc
+
+	activeTasks *common.SharedGauge
+
+	wg        sync.WaitGroup
+	chans     []chan Batch
+	triggerCh chan chan struct{}
+}
+
+type SchedulerOptions struct {
+	Logger logrus.FieldLogger
+	// Number of workers to process tasks. Defaults to the number of CPUs - 1.
+	Workers int
+	// The interval at which the scheduler checks the queues for tasks. Defaults to 1 second.
+	ScheduleInterval time.Duration
+	// Function to be called when the scheduler is closed
+	OnClose func()
+}
+
+func NewScheduler(opts SchedulerOptions) *Scheduler {
+	if opts.Logger == nil {
+		opts.Logger = logrus.New()
+	}
+	opts.Logger = opts.Logger.WithField("component", "queue-scheduler")
+
+	if opts.Workers <= 0 {
+		opts.Workers = max(1, runtime.GOMAXPROCS(0)-1)
+	}
+
+	if opts.ScheduleInterval == 0 {
+		it, err := time.ParseDuration(os.Getenv("QUEUE_SCHEDULER_INTERVAL"))
+		if err != nil {
+			it = 1 * time.Second
```

Comment:
minor: it may be worth at least logging about a potential invalid value set in the env variable?